File: archive/z/zig/base64-encode-decode.zig
const std = @import("std");
const decoder = std.base64.standard.Decoder;
const encoder = std.base64.standard.Encoder;
const process = std.process;

const stdout = std.io.getStdOut().writer();

fn die() !void {
    try stdout.writeAll("Usage: please provide a mode and a string to encode/decode\n");
    process.exit(1);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var args = process.args();

    _ = args.next();
    const mode_opt = args.next();
    const src_opt = args.next();

    if (mode_opt == null or src_opt == null) {
        try die();
    }

    const mode = mode_opt.?;
    const src = src_opt.?;

    if (src.len == 0) {
        try die();
    }

    if (std.mem.eql(u8, mode, "encode")) {
        const buffer = try allocator.alloc(u8, encoder.calcSize(src.len));
        defer allocator.free(buffer);

        _ = encoder.encode(buffer, src);
        try stdout.writeAll(buffer);
        _ = try stdout.write("\n");
    } else if (std.mem.eql(u8, mode, "decode")) {
        errdefer {
            die() catch unreachable;
        }
        const buflen = try decoder.calcSizeForSlice(src);

        const buffer = try allocator.alloc(u8, buflen);
        defer allocator.free(buffer);

        decoder.decode(buffer, src) catch {
            try die();
        };

        try stdout.writeAll(buffer);
        _ = try stdout.write("\n");
    } else {
        try die();
    }
}

File: archive/g/go/base64-encode-decode.go
package main

import (
	"encoding/base64"
	"fmt"
	"os"
)

func die() {
	fmt.Println("Usage: please provide a mode and a string to encode/decode")
	os.Exit(1)
}

func main() {
	if len(os.Args) < 3 {
		die()
	}

	enc := base64.StdEncoding

	if len(os.Args[2]) == 0 {
		die()
	}

	switch os.Args[1] {
	case "encode":
		fmt.Println(enc.EncodeToString([]byte(os.Args[2])))
		return
	case "decode":
		s, err := enc.DecodeString(os.Args[2])
		if err != nil {
			die()
		}

		fmt.Println(string(s))
	default:
		die()
	}
}
